### PR TITLE
authz: Use conf.Watch in SubRepoPermsClient

### PR DESCRIPTION
**Pull request by ryanslade**
Instead of running conf.Get every time we call client.Enabled, which can
be very frequent, we use conf.Watch to update an atomic bool indicating
whether we're enabled.

Benchmark comparison of FilterActorPaths:

```
name                 old time/op    new time/op    delta
FilterActorPaths-10    2.51ms ± 1%    2.13ms ± 0%  -15.27%  (p=0.000 n=10+10)

name                 old paths/s    new paths/s    delta
FilterActorPaths-10     1.99M ± 1%     2.35M ± 0%  +18.03%  (p=0.000 n=10+10)
```

Part of https://github.com/sourcegraph/sourcegraph/issues/41347

## Test plan

All tests pass

### CODE REVIEW

**Review by sashaostrikov — APPROVED** [review]
LGTM!

**Review by keegancsmith — APPROVED** [review]
I see this constructor is only called once per process and we are _already_ using conf.Watch in the constructor. This is an anti-pattern, but is no worse than the current usage and this is an improvement. So approving :)

File: internal/authz/sub_repo_perms.go [hunk 1]
```diff
@@ -128,10 +130,19 @@ func NewSubRepoPermsClient(permissionsGetter SubRepoPermissionsGetter) (*SubRepo
 		return nil, errors.Wrap(err, "creating LRU cache")
 	}
 
+	enabled := atomic.NewBool(false)
+
 	conf.Watch(func() {
-		if c := conf.Get(); c.ExperimentalFeatures != nil && c.ExperimentalFeatures.SubRepoPermissions != nil && c.ExperimentalFeatures.SubRepoPermissions.UserCacheSize > 0 {
+		c := conf.Get()
+		if c.ExperimentalFeatures != nil && c.ExperimentalFeatures.SubRepoPermissions != nil && c.ExperimentalFeatures.SubRepoPermissions.UserCacheSize > 0 {
```
Comment (keegancsmith) [inline internal/authz/sub_repo_perms.go:137]: unrelated: if UserCacheSize becomes unset / zero we don't reset the size back to the default.

**Comment by ryanslade** [comment]
@keegancsmith Cool, thanks for the tips. I ran it 10 times and updated the PR comment

**Comment by ryanslade** [comment]
PTAL, we now reset the cache size correctly

**Pull request merged**